gatsby/content/docs/2020-04-14-stateresv2.mdx

```diff
@@ -0,0 +1,329 @@
+
+To update the partial state, start at the oldest event and loop through the list, authing each event against the current partial state. If the event passes auth checks then update the partial state to include the event before moving onto the next iteration:
+
+```
+for each sorted_control_events() as e:
+    if auth_against_partial_state(e):
+        partial_state[e.event_type()][e.state_key()] = e
+```
+
+Events that fail the auth check should be ignored and should not find their way into the partial state.
+
+### Creating the power level mainline
+
+Once the control events have been applied to the partial state, the remainder of the normal state events should be ordered and processed based on the ordering of their power level auth events. This is done using a technique called _mainline ordering_.
+
+The _power level_ _mainline_ is created by following the path of power level events, starting from the resolved `m.room.power_level` event from the partial state, back to the point that the room was created:
+
+```
+var resolved_power_level_event: partial_state["m.room.power_level"][""] as event
+var power_level_mainline:       list(event)
+
+func mainline_iterate(e as event):
+    power_level_mainline += e
+    for each e.auth_events as auth_event:
+        if auth_event.event_type == "m.room.power_level":
+            mainline_iterate(auth_event)
+```
+
+We refer to the time period between any two given power level events as an _epoch_.
+
+In the below illustration, the "resolved" event is the event that appears in the partial state from the previous section.
+
+![drawing showing partial state example](/docs/guides/img/2020-04-14-stateresv2-2.svg)
+
+### Mainline ordering of normal state events
+
+Since the mainline will be ordered based on when the power level changes happened, and other state events use power level events as auth events, it is possible to sort the order of normal state events based on which mainline power level event they most closely refer to, or by definition, which epoch the event happened in.
+
+Note that events may appear that refer to power level events which don't appear in the mainline, in which case it is necessary to step through the power level events iteratively until you encounter one that is on the mainline. This is referred to as the _closest mainline event_:
+
+```
+func is_in_mainline(e as event) returns boolean:
+    if e in power_level_mainline:
+        return true
+    else:
+        return false
+
+func get_closest_mainline_event(e as event) returns event:
+    var closest_mainline_event: event
+
+    func closest_iterate(e as event):
+    if is_in_mainline(e):
+            closest_mainline_event = e
+        else:
+            for each e.auth_events as auth_event:
+                if auth_event.event_type == "m.room.power_level":
+                    closest_iterate(auth_event)
+
+    closest_iterate(e)
+    return closest_mainline_event
```

Comment:
I feel like you don't need `closest_iterate` and you could just make `get_closest_mainline_event` a recursive function. Is the reason you make an extra function because you don't want to run `closest_mainline_event = event` over and over?

Reply:
I just felt like this was a clearer way of demonstrating the bounds of recursion, and because it's pretty much how it was implemented in gomatrixserverlib.

Reply:
Fair enough, I agree that it's pretty clear code :)